File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
     int[] nums; 
        
     public TreeNode helper(int l ,int r){
        if(l> r) return null ; 

        int p = (l+r) /2 ; 
        if(( l+r) % 2==1 ) p++; 
        
        TreeNode root = new TreeNode(nums[p]) ;
        root.left = helper(l , p-1) ; 
        root.right = helper(p+1 , r) ; 
        return root ; 
    
    }
    
    
    public TreeNode sortedArrayToBST(int[] nums) {
        
        this.nums = nums ; 
        return helper( 0 , nums.length -1 ) ; 
    }
}

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& num) {
        

        if(num.size() ==0) return NULL ; 
        if(num.size() ==1) return new TreeNode(num[0]);

        int m = num.size()/2 ;
        TreeNode* root = new TreeNode(num[m]); 

        vector<int> Left(num.begin() , num.begin()+m); 
        vector<int> Right(num.begin()+m+1 , num.end()); 

        root->left = sortedArrayToBST(Left) ; 
        root->right = sortedArrayToBST(Right); 

        return root ; 
        
        
    }
};